app/services/ui_notifications.rb

```diff
@@ -0,0 +1,37 @@
+# Namespace for Foreman UI notifications event handling
+module UINotifications
+  class Base
+    attr_reader :subject, :event
+
+    def self.deliver!(subject, event)
+      new(subject, event).deliver!
+    end
+
+    def initialize(subject, event)
+      raise(Foreman::Exception, 'must provide notification subject') if subject.nil?
+      raise(Foreman::Exception, 'invalid event') unless allowed_events.include?(event)
+      @subject = subject
+      @event = event
+    end
+
+    def deliver!
+      logger.debug("Notification event: #{event} - checking for notifications")
+      send(event) if respond_to?(event, true)
```

Comment:
Nitpick, `public_send`, or even better, the whole line could just be `try(event)`, correct?